File: src/lib/types/flashcard.ts
export interface Flashcard {
  id: number;
  textId: number;
  userId: number;
  originalText: string;
  clozeText: string;
  clozeIndex: number;
  displayIndex: number;
  clozeNumber: number;
  createdAt: string;
  updatedAt: string;
  clozeNoteId: number | null;
  due: string;
  stability: number;
  difficulty: number;
  elapsedDays: number;
  scheduledDays: number;
  reps: number;
  lapses: number;
  state: number;
  lastReview: string | null;
}

export interface ClozeNote {
  id: number;
  textId: number;
  userId: number;
  originalText: string;
  parsedSegments: string;
  clozeCount: number;
  createdAt: string;
  updatedAt: string;
}

export interface ClozeSegment {
  clozeNumber: number;
  text: string;
  hint: string | null;
  startPosition: number;
  endPosition: number;
}

export interface CreateFlashcardRequest {
  textId: number;
  selectedText: string;
  clozeText: string;
}

export interface ClozeData {
  fullText: string;
  clozedWord: string;
  startIndex: number;
  endIndex: number;
  contextBefore: string;
  contextAfter: string;
}

export interface ReviewResult {
  flashcardId: string;
  quality: number;
  timestamp: Date;
  responseTime: number;
}

export type ReviewQuality = 0 | 1 | 2 | 3 | 4 | 5;

export interface FlashcardPreview {
  html: string;
  clozeNumber: number;
}

export type ReviewFilter =
  | { type: 'global' }
  | { type: 'text'; textId: number }
  | { type: 'folder'; folderId: string };

File: src-tauri/src/models/flashcard.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use sqlx::FromRow;

#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
#[serde(rename_all = "camelCase")]
pub struct Flashcard {
    pub id: i64,
    pub text_id: i64,
    pub user_id: i64,
    pub original_text: String,
    pub cloze_text: String,
    pub cloze_index: i64,
    pub display_index: i64,
    pub cloze_number: i64,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
    pub cloze_note_id: Option<i64>,
    pub due: DateTime<Utc>,
    pub stability: f64,
    pub difficulty: f64,
    pub elapsed_days: i64,
    pub scheduled_days: i64,
    pub reps: i64,
    pub lapses: i64,
    pub state: i64,
    pub last_review: Option<DateTime<Utc>>,
}
